common/datavalues/src/data_array_logic.rs

```diff
@@ -40,4 +37,29 @@ impl DataArrayLogic {
             )))
         }
     }
+
+    fn data_array_negation(val: &DataColumnarValue) -> Result<DataArrayRef> {
+        match val {
+            DataColumnarValue::Array(array) => {
+                let arr = downcast_array!(array, BooleanArray)?;
+                Ok(Arc::new(
+                    common_arrow::arrow::compute::not(arr).map_err(ErrorCodes::from)?
+                ))
+            }
+            _ => Result::Err(ErrorCodes::BadDataValueType(format!(
+                "DataValue Error: Cannot do data_array Not, array:{:?}",
```

Comment:
I think DataColumnarValue::Constant case should be considered.


1. Using `val.to_array()` , return the value as `DataColumnarValue::Array` 
2. Creating a array with just one element, and apply the `compute::not`, then return the value as `DataColumnarValue::Constant`

The second option can have better performance.

Reply:
We can have the stateless tests like `select number from numbers(10) where not ( 5 )`

Reply:
Thanks for the suggestions. The function returns Result<DataArrayRef>, is it possible to return value as DataColumnarValue::Constant?

Reply:
It's ok.

Reply:
I have changed the code. Could you review it again?